pkg/terraform/controller.go

```diff
@@ -0,0 +1,201 @@
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/retry"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	xpmeta "github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/ratelimiter"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	xpresource "github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	"github.com/crossplane-contrib/terrajet/pkg/conversion"
+	"github.com/crossplane-contrib/terrajet/pkg/meta"
+	"github.com/crossplane-contrib/terrajet/pkg/terraform/resource"
+)
+
+const (
+	errUnexpectedObject = "the managed resource is not an Terraformed resource"
+)
+
+// ProviderConfigFn is a function that returns provider specific configuration
+// like provider credentials used to connect to cloud APIs.
+type ProviderConfigFn func(ctx context.Context, client client.Client, mg xpresource.Managed) ([]byte, error)
+
+// SetupController setups controller for a Terraform managed resource
+func SetupController(mgr ctrl.Manager, l logging.Logger, obj client.Object, of schema.GroupVersionKind, pcFn ProviderConfigFn) error {
+	name := managed.ControllerName(of.GroupKind().String())
+
+	rl := ratelimiter.NewDefaultProviderRateLimiter(ratelimiter.DefaultProviderRPS)
+	o := controller.Options{
+		RateLimiter: ratelimiter.NewDefaultManagedRateLimiter(rl),
+	}
+
+	r := managed.NewReconciler(mgr,
+		xpresource.ManagedKind(of),
+		managed.WithInitializers(),
+		managed.WithExternalConnecter(&connector{kube: mgr.GetClient(), providerConfig: pcFn, logger: l}),
+		managed.WithLogger(l.WithValues("controller", name)),
+		managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name))))
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithOptions(o).
+		For(obj).
+		Complete(r)
+}
+
+type connector struct {
+	kube           client.Client
+	providerConfig ProviderConfigFn
+	logger         logging.Logger
+}
+
+func (c *connector) Connect(ctx context.Context, mg xpresource.Managed) (managed.ExternalClient, error) {
+	tr, ok := mg.(resource.Terraformed)
+	if !ok {
+		return nil, errors.New(errUnexpectedObject)
+	}
+
+	// TODO(hasan): create and pass the implementation of tfcli builder once available
+	/*
+		pc, err := c.providerConfig(ctx, c.kube, mg)
+		if err != nil {
+			return nil, errors.Wrap(err, "cannot get provider config")
+		}
+		tfcb := tfcli.NewClientBuilder().
+			WithLogger(c.logger).
+			WithResourceName(tr.GetName()).
+			WithHandle(string(tr.GetUID())).
+			WithProviderConfiguration(pc).
+			WithResourceType(tr.GetTerraformResourceType())
+
+		tfcli, err := conversion.BuildClientForResource(tfcb, tr)
+	*/
+
+	tfcli, err := conversion.BuildClientForResource(nil, tr)
```

Comment:
We may operate under the assumption that workspace initialization is not costly (because we have the means to do that by avoiding provider plugin downloads each time a workspace is initialized) but I do not like the idea of operating under extra assumptions. For example, development would require special setup & configuration on a developer's machine.